File: evaluations/gridcal/tests/expressiveness/test_a3_dcopf.py
"""
Test A-3: Solve DC OPF with differentiated gen costs and 70% branch derating on TINY.

Dimension: expressiveness
Network: TINY (IEEE 39-bus New England)
Pass condition: Converges. Optimal dispatch and LMPs/shadow prices extractable. With
    differentiated costs and 70% derating, at least 2 branches have non-zero shadow prices
    (binding flow constraints). Report max LMP spread across buses.
Tool: gridcal (VeraGridEngine) 5.6.28
Solver: HiGHS
"""

from __future__ import annotations

import csv
import json
import sys
import time
import traceback
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent.parent / "shared"))

from matpower_loader import load_gridcal

# Cost mapping from gen_temporal_params.csv tech_class_key
COST_MAP = {
    "hydro": {"c1": 5.0, "c2": 0.005},
    "nuclear": {"c1": 10.0, "c2": 0.010},
    "coal_large": {"c1": 25.0, "c2": 0.025},
    "gas_CC": {"c1": 40.0, "c2": 0.040},
}

BRANCH_DERATING = 0.70


def run(
    network_file: str = "data/networks/case39.m",
    timeseries_dir: str | None = "data/timeseries/case39",
) -> dict:
    """Execute A-3 DCOPF test and return structured results."""
    results = {
        "status": "fail",
        "wall_clock_seconds": 0.0,
        "details": {},
        "errors": [],
        "workarounds": [],
    }

    start = time.perf_counter()
    try:
        import VeraGridEngine as vge
        from VeraGridEngine.enumerations import MIPSolvers, SolverType

        # 1. Load network
        grid = load_gridcal(network_file)
        n_buses = grid.get_bus_number()
        generators = grid.get_generators()
        branches = grid.get_branches()
        results["details"]["bus_count"] = n_buses
        results["details"]["gen_count"] = len(generators)
        results["details"]["branch_count"] = len(branches)

        # 2. Load augmented cost data from timeseries directory
        ts_dir = Path(timeseries_dir) if timeseries_dir else None
        if ts_dir and (ts_dir / "gen_temporal_params.csv").exists():
            gen_params = {}
            with open(ts_dir / "gen_temporal_params.csv") as f:
                reader = csv.DictReader(f)
                for row in reader:
                    gen_idx = int(row["gen_index"])
                    gen_params[gen_idx] = row

            # Apply differentiated costs to generators
            for idx, gen in enumerate(generators):
                if idx in gen_params:
                    tech_key = gen_params[idx]["tech_class_key"]
                    if tech_key in COST_MAP:
                        gen.Cost = COST_MAP[tech_key]["c1"]  # linear cost $/MWh
                        gen.Cost2 = COST_MAP[tech_key]["c2"]  # quadratic cost $/MW^2h
                        gen.Cost0 = 0.0  # constant cost

            results["details"]["cost_augmentation"] = (
                "Applied differentiated costs from gen_temporal_params.csv"
            )
            results["details"]["cost_mapping"] = {k: v["c1"] for k, v in COST_MAP.items()}
        else:
            results["errors"].append("gen_temporal_params.csv not found; using default costs")

        # 3. Apply 70% branch derating
        for branch in branches:
            if hasattr(branch, "rate") and branch.rate > 0:
                branch.rate = branch.rate * BRANCH_DERATING
            elif hasattr(branch, "Rate") and branch.Rate > 0:
                branch.Rate = branch.Rate * BRANCH_DERATING
        results["details"]["branch_derating"] = BRANCH_DERATING

        # 4. Configure and solve DC OPF
        opf_opts = vge.OptimalPowerFlowOptions(
            solver=SolverType.LINEAR_OPF,
            mip_solver=MIPSolvers.HIGHS,
        )

        opf_results = vge.linear_opf(grid, opf_opts)

        # 5. Extract results
        converged = bool(opf_results.converged)
        results["details"]["converged"] = converged
        results["details"]["solver"] = "HiGHS"

        if not converged:
            results["errors"].append("DC OPF did not converge")
            return results

        # Generator dispatch
        gen_power = opf_results.generator_power
        gen_names = [g.name for g in generators]
        results["details"]["generator_dispatch_mw"] = {
            gen_names[i]: float(gen_power[i]) for i in range(len(gen_power))
        }
        results["details"]["total_generation_mw"] = float(np.sum(gen_power))

        # LMPs / bus shadow prices
        lmps = opf_results.bus_shadow_prices
        bus_names = [b.name for b in grid.get_buses()]

        if lmps is not None and len(lmps) > 0:
            results["details"]["lmps"] = {bus_names[i]: float(lmps[i]) for i in range(len(lmps))}
            lmp_min = float(np.min(lmps))
            lmp_max = float(np.max(lmps))
            lmp_spread = lmp_max - lmp_min
            results["details"]["lmp_min"] = lmp_min
            results["details"]["lmp_max"] = lmp_max
            results["details"]["lmp_spread"] = lmp_spread
            results["details"]["lmp_mean"] = float(np.mean(lmps))
        else:
            results["details"]["lmp_spread"] = 0.0
            results["errors"].append("No LMP/shadow price data available")

        # Branch flows and shadow prices
        sf = opf_results.Sf
        loading = opf_results.loading
        branch_names_all = [b.name for b in branches]

        results["details"]["branch_flows_mw"] = {
            branch_names_all[i]: float(np.real(sf[i])) for i in range(len(sf))
        }
        results["details"]["branch_loading_pct"] = {
            branch_names_all[i]: float(np.abs(loading[i]) * 100) for i in range(len(loading))
        }

        # Count binding branches (loading >= 99%)
        binding_branches = []
        for i in range(len(loading)):
            if np.abs(loading[i]) >= 0.99:
                binding_branches.append(branch_names_all[i])
        results["details"]["binding_branch_count"] = len(binding_branches)
        results["details"]["binding_branches"] = binding_branches

        # Check for branch overloads/shadow prices
        if hasattr(opf_results, "overloads") and opf_results.overloads is not None:
            overloads = opf_results.overloads
            nonzero_shadow = np.sum(np.abs(overloads) > 1e-6)
            results["details"]["branches_with_shadow_prices"] = int(nonzero_shadow)
            results["details"]["overloads"] = {
                branch_names_all[i]: float(overloads[i])
                for i in range(len(overloads))
                if abs(overloads[i]) > 1e-6
            }

        # 6. Check max branch loading for soft constraint detection (v11)
        max_loading_abs = float(np.max(np.abs(loading)))
        results["details"]["max_branch_loading"] = max_loading_abs
        results["details"]["max_branch_loading_pct"] = f"{max_loading_abs * 100:.2f}%"

        # Find which branch has max loading
        max_idx = int(np.argmax(np.abs(loading)))
        results["details"]["max_loading_branch"] = branch_names_all[max_idx]

        # Soft constraint detection: if any branch exceeds 100% + 1e-4 tolerance
        soft_constraint_detected = max_loading_abs > (1.0 + 1e-4)
        results["details"]["soft_constraint_detected"] = soft_constraint_detected

        if soft_constraint_detected:
            results["details"]["soft_constraint_note"] = (
                f"Branch {branch_names_all[max_idx]} has loading "
                f"{max_loading_abs * 100:.2f}%, exceeding 100% + 1e-4 tolerance. "
                "GridCal linear_opf uses soft branch flow constraints (LP slack variables). "
                "[tool-specific: soft constraint formulation in linear_opf]"
            )

        # 7. Check pass condition
        pass_checks = {
            "converged": converged,
            "lmps_extractable": lmps is not None and len(lmps) > 0,
            "binding_branches_ge_2": len(binding_branches) >= 2,
        }
        results["details"]["pass_checks"] = pass_checks

        if all(pass_checks.values()):
            if soft_constraint_detected:
                # v11: soft constraints -> partial_pass
                results["status"] = "partial_pass"
                results["workarounds"].append(
                    "GridCal linear_opf uses soft branch flow constraints. "
                    f"Max loading {max_loading_abs * 100:.2f}% exceeds hard limit. "
                    "This is a legitimate numerical stabilization technique but "
                    "must be labeled as soft-constraint DCOPF."
                )
            else:
                results["status"] = "pass"
        else:
            failing = [k for k, v in pass_checks.items() if not v]
            results["errors"].append(f"Failed checks: {failing}")

    except Exception as e:
        results["errors"].append(f"{type(e).__name__}: {e}")
        results["details"]["traceback"] = traceback.format_exc()
    finally:
        results["wall_clock_seconds"] = time.perf_counter() - start

    return results


if __name__ == "__main__":
    result = run()
    print(json.dumps(result, indent=2, default=str))

File: evaluations/powersimulations/tests/expressiveness/test_a3_dcopf.jl
#=
Test A-3: DCOPF with differentiated costs and 70% branch derating

Dimension: expressiveness
Network: TINY (IEEE 39-bus)
Pass condition (v11): Converges. Optimal dispatch and LMPs extractable. With differentiated
  costs and 70% derating, >=2 branches have non-zero shadow prices. Report max LMP spread.
  v11 requires hard constraint enforcement: max(loading_percent) <= 100 + 1e-4 p.u.
  Soft-constraint DCOPF = partial_pass.
Tool: PowerSimulations.jl v0.30.2
=#

using PowerSystems
using PowerSimulations
using HiGHS
using JuMP
using JSON
using Logging
using DataFrames
using CSV
using Dates
using TimeSeries: TimeArray

# Suppress verbose logging
global_logger(ConsoleLogger(stderr, Logging.Error))

function peak_rss_mb()
    for line in eachline("/proc/self/status")
        if startswith(line, "VmHWM:")
            return parse(Float64, split(line)[2]) / 1024  # kB to MB
        end
    end
    return nothing
end

# Cost mapping from README: tech_class_key -> c1 ($/MWh)
const COST_MAP = Dict("hydro" => 5.0, "nuclear" => 10.0, "coal_large" => 25.0, "gas_CC" => 40.0)

function setup_system(network_file, timeseries_dir)
    sys = System(network_file)

    # Apply differentiated costs
    params = CSV.read(joinpath(timeseries_dir, "gen_temporal_params.csv"), DataFrame)
    cost_changes = Dict{String,Any}[]
    for row in eachrow(params)
        c1 = get(COST_MAP, row.tech_class_key, 30.0)
        c2 = c1 * 0.001
        for gen in get_components(ThermalStandard, sys)
            if get_number(get_bus(gen)) == row.bus_id
                set_operation_cost!(
                    gen,
                    ThermalGenerationCost(CostCurve(QuadraticCurve(c2, c1, 0.0)), 0.0, 0.0, 0.0),
                )
                push!(
                    cost_changes,
                    Dict(
                        "gen" => get_name(gen),
                        "bus" => row.bus_id,
                        "tech" => row.tech_class_key,
                        "c1" => c1,
                        "c2" => c2,
                    ),
                )
                break
            end
        end
    end

    # Apply 70% branch derating — store original ratings for loading calculation
    derated = 0
    for line in get_components(Line, sys)
        set_rating!(line, get_rating(line) * 0.7)
        derated += 1
    end
    for xfmr in get_components(Transformer2W, sys)
        set_rating!(xfmr, get_rating(xfmr) * 0.7)
        derated += 1
    end
    for xfmr in get_components(TapTransformer, sys)
        set_rating!(xfmr, get_rating(xfmr) * 0.7)
        derated += 1
    end

    # Add load time series (required by PowerSimulations)
    timestamps = [DateTime("2024-01-01"), DateTime("2024-01-01") + Hour(1)]
    for load in get_components(PowerLoad, sys)
        add_time_series!(
            sys, load, SingleTimeSeries("max_active_power", TimeArray(timestamps, [1.0, 1.0]))
        )
    end
    transform_single_time_series!(sys, Hour(1), Hour(1))

    return sys, cost_changes, derated
end

function build_and_solve(sys, solver)
    # DCPPowerModel with NodalBalanceActiveConstraint duals for bus-level LMPs
    template = ProblemTemplate(NetworkModel(DCPPowerModel; duals=[NodalBalanceActiveConstraint]))
    set_device_model!(template, ThermalStandard, ThermalDispatchNoMin)
    set_device_model!(template, PowerLoad, StaticPowerLoad)
    set_device_model!(template, Line, StaticBranch)
    set_device_model!(template, Transformer2W, StaticBranch)
    set_device_model!(template, TapTransformer, StaticBranch)

    model = DecisionModel(template, sys; optimizer=solver)
    build!(model; output_dir=mktempdir())
    solve!(model)
    return model
end

function run(
    network_file::String="/workspace/data/networks/case39.m";
    timeseries_dir::Union{String,Nothing}="/workspace/data/timeseries/case39",
)
    results = Dict(
        "status" => "fail",
        "wall_clock_seconds" => 0.0,
        "details" => Dict(),
        "errors" => String[],
        "workarounds" => String[],
    )

    try
        if timeseries_dir === nothing
            push!(results["errors"], "timeseries_dir required for A-3")
            return results
        end

        solver = optimizer_with_attributes(
            HiGHS.Optimizer,
            "time_limit" => 300.0,
            "presolve" => "on",
            "threads" => 1,
            "output_flag" => false,
        )

        # Warm-up (JIT compilation)
        sys_w, _, _ = setup_system(network_file, timeseries_dir)
        build_and_solve(sys_w, solver)

        # Timed run
        sys, cost_changes, derated = setup_system(network_file, timeseries_dir)
        base_power = get_base_power(sys)

        t0 = time()
        model = build_and_solve(sys, solver)
        elapsed = time() - t0

        results["wall_clock_seconds"] = elapsed
        results["details"]["peak_memory_mb"] = peak_rss_mb()
        results["details"]["base_power_mva"] = base_power
        results["details"]["cost_changes"] = cost_changes
        results["details"]["branches_derated"] = derated

        # Workaround: PSI requires time series even for single-snapshot OPF
        push!(
            results["workarounds"],
            "PowerSimulations.jl requires deterministic time series data even for single-snapshot DCOPF. Added 1-step forecast with multiplier=1.0 for all loads.",
        )

        # Extract results
        res = OptimizationProblemResults(model)

        # Dispatch (returned in MW by PSI)
        dispatch_df = read_variable(res, "ActivePowerVariable__ThermalStandard")
        dispatch_summary = Dict{String,Any}()
        for col in names(dispatch_df)
            col == "DateTime" && continue
            dispatch_summary[col] = Dict("dispatch_mw" => dispatch_df[1, col])
        end
        results["details"]["dispatch"] = dispatch_summary

        # LMPs from nodal balance duals
        nodal_dual = read_dual(res, "NodalBalanceActiveConstraint__ACBus")
        lmp_data = Dict{String,Float64}()
        lmp_vals = Float64[]
        for col in names(nodal_dual)
            col == "DateTime" && continue
            raw_dual = nodal_dual[1, col]
            lmp_mwh = -raw_dual / base_power  # convert to $/MWh
            lmp_data[col] = lmp_mwh
            push!(lmp_vals, lmp_mwh)
        end

        lmp_min = minimum(lmp_vals)
        lmp_max = maximum(lmp_vals)
        lmp_spread = lmp_max - lmp_min

        results["details"]["lmps"] = lmp_data
        results["details"]["lmp_summary"] = Dict(
            "min_dollar_per_mwh" => round(lmp_min; digits=2),
            "max_dollar_per_mwh" => round(lmp_max; digits=2),
            "spread_dollar_per_mwh" => round(lmp_spread; digits=2),
            "mean_dollar_per_mwh" => round(sum(lmp_vals) / length(lmp_vals); digits=2),
            "n_buses" => length(lmp_vals),
        )

        # ===== v11: Hard constraint enforcement verification =====
        # Check ALL branch types (Line, Transformer2W, TapTransformer) for loading
        all_loading = Float64[]
        binding_branches = Dict{String,Any}[]

        # Line flows
        flows_line = read_variable(res, "FlowActivePowerVariable__Line")
        for line in get_components(Line, sys)
            ln = get_name(line)
            if ln in names(flows_line)
                flow_mw = abs(flows_line[1, ln])
                rating_mw = get_rating(line) * base_power
                if rating_mw > 0
                    loading_pct = flow_mw / rating_mw * 100.0
                    push!(all_loading, loading_pct)
                    if loading_pct > 99.0
                        push!(
                            binding_branches,
                            Dict(
                                "branch" => ln,
                                "type" => "Line",
                                "flow_mw" => round(flow_mw; digits=1),
                                "rating_mw" => round(rating_mw; digits=1),
                                "loading_pct" => round(loading_pct; digits=2),
                            ),
                        )
                    end
                end
            end
        end

        # Check Transformer2W flows if variable exists
        try
            flows_t2w = read_variable(res, "FlowActivePowerVariable__Transformer2W")
            for xfmr in get_components(Transformer2W, sys)
                xn = get_name(xfmr)
                if xn in names(flows_t2w)
                    flow_mw = abs(flows_t2w[1, xn])
                    rating_mw = get_rating(xfmr) * base_power
                    if rating_mw > 0
                        loading_pct = flow_mw / rating_mw * 100.0
                        push!(all_loading, loading_pct)
                        if loading_pct > 99.0
                            push!(
                                binding_branches,
                                Dict(
                                    "branch" => xn,
                                    "type" => "Transformer2W",
                                    "flow_mw" => round(flow_mw; digits=1),
                                    "rating_mw" => round(rating_mw; digits=1),
                                    "loading_pct" => round(loading_pct; digits=2),
                                ),
                            )
                        end
                    end
                end
            end
        catch
            # Variable may not exist if no Transformer2W in system
        end

        # Check TapTransformer flows if variable exists
        try
            flows_tap = read_variable(res, "FlowActivePowerVariable__TapTransformer")
            for xfmr in get_components(TapTransformer, sys)
                xn = get_name(xfmr)
                if xn in names(flows_tap)
                    flow_mw = abs(flows_tap[1, xn])
                    rating_mw = get_rating(xfmr) * base_power
                    if rating_mw > 0
                        loading_pct = flow_mw / rating_mw * 100.0
                        push!(all_loading, loading_pct)
                        if loading_pct > 99.0
                            push!(
                                binding_branches,
                                Dict(
                                    "branch" => xn,
                                    "type" => "TapTransformer",
                                    "flow_mw" => round(flow_mw; digits=1),
                                    "rating_mw" => round(rating_mw; digits=1),
                                    "loading_pct" => round(loading_pct; digits=2),
                                ),
                            )
                        end
                    end
                end
            end
        catch
            # Variable may not exist if no TapTransformer in system
        end

        results["details"]["binding_branches"] = binding_branches
        results["details"]["num_binding_branches"] = length(binding_branches)

        # v11 hard constraint verification
        max_loading = isempty(all_loading) ? 0.0 : maximum(all_loading)
        hard_constraint_threshold = 100.0 + 1e-2  # 100% + 1e-4 p.u. expressed as percentage
        hard_constraint_met = max_loading <= hard_constraint_threshold

        results["details"]["hard_constraint_enforcement"] = Dict(
            "max_loading_pct" => round(max_loading; digits=4),
            "threshold_pct" => hard_constraint_threshold,
            "hard_constraints_enforced" => hard_constraint_met,
            "total_branches_checked" => length(all_loading),
            "note" => "v11: max(loading_percent) must be <= 100 + 1e-4 p.u. for hard constraint pass",
        )

        # Check for soft constraints: inspect if slack variables are present in the JuMP model
        oc = PowerSimulations.get_optimization_container(model)
        jm = PowerSimulations.get_jump_model(oc)
        has_flow_slacks = false
        try
            # Check if flow slack variables exist in the model
            all_var_names = [
                string(JuMP.name(v)) for v in JuMP.all_variables(jm) if JuMP.name(v) != ""
            ]
            flow_slack_vars = filter(
                n -> occursin("Slack", n) && occursin("Flow", n), all_var_names
            )
            has_flow_slacks = !isempty(flow_slack_vars)
            results["details"]["hard_constraint_enforcement"]["flow_slack_variables_found"] =
                flow_slack_vars
        catch e
            results["details"]["hard_constraint_enforcement"]["slack_check_error"] = string(e)
        end

        # Objective and termination
        results["details"]["objective_value"] = objective_value(jm)
        results["details"]["termination_status"] = string(termination_status(jm))

        # Production costs
        try
            cost_df = read_expression(res, "ProductionCostExpression__ThermalStandard")
            cost_summary = Dict{String,Any}()
            total_cost = 0.0
            for col in names(cost_df)
                col == "DateTime" && continue
                val = cost_df[1, col]
                cost_summary[col] = round(val; digits=2)
                total_cost += val
            end
            results["details"]["production_costs"] = cost_summary
            results["details"]["total_production_cost"] = round(total_cost; digits=2)
        catch e
            results["details"]["production_cost_error"] = string(e)
        end

        # Pass condition checks
        has_dispatch = !isempty(dispatch_summary)
        has_lmps = !isempty(lmp_data)
        shadow_check = length(binding_branches) >= 2

        results["details"]["pass_checks"] = Dict(
            "converged" => true,
            "has_dispatch" => has_dispatch,
            "has_lmps" => has_lmps,
            "num_binding_branches" => length(binding_branches),
            "binding_branches_gte_2" => shadow_check,
            "lmp_spread_dollar_per_mwh" => round(lmp_spread; digits=2),
            "hard_constraints_enforced" => hard_constraint_met,
            "uses_soft_constraints" => has_flow_slacks,
        )

        if has_dispatch && has_lmps && shadow_check
            if has_flow_slacks && !hard_constraint_met
                # Soft constraints detected and limits exceeded
                results["status"] = "partial_pass"
                push!(
                    results["workarounds"],
                    "Soft branch flow constraints detected (flow slack variables present). max_loading=$(round(max_loading; digits=2))% exceeds hard constraint threshold.",
                )
            else
                results["status"] = "pass"
            end
        elseif has_dispatch && has_lmps
            results["status"] = "qualified_pass"
            push!(
                results["workarounds"],
                "Only $(length(binding_branches)) binding branches found (need >= 2).",
            )
        else
            push!(results["errors"], "Pass condition not met")
        end

    catch e
        push!(results["errors"], string(typeof(e), ": ", sprint(showerror, e)))
        results["details"]["traceback"] = sprint(showerror, e, catch_backtrace())
    end

    return results
end

if abspath(PROGRAM_FILE) == @__FILE__
    result = run()
    println(JSON.json(result, 2))
end
